enhancements/authentication/direct-oidc-configuration.md

```diff
@@ -0,0 +1,463 @@
+> TODO: will this break the Projects API?
+
+The platform MUST provide means of administrative user authentication even without
+the oauth-server. In case of standalone OpenShift, this is provided by the
+`system:admin` certificate/key pair present in the installation kubeconfig.
+
+Once `Type: OIDC` is set and the configuration passes validation, the operator
+configuring the kube-apiserver converts the API values into appropriate flags.
+In later versions, when the underlying kube-apiserver implements [a stable structured
+authentication configuration](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3331-structured-authentication-configuration),
+the values from the API get transformed into this new structure.
+
+#### Hypershift
+
+The `Authentication` API that's being modified in this enhancement is already
+vendored into Hypershift's `ClusterConfiguration` type, the API changes therefore
+carry over to Hypershift cleanly.
+
+Hypershift is not prepared today for Tech Preview features and does not appear
+to have a way to mark cluster support to be limited as such. However, this feature
+in its first iteration will not be suitable for production.
+
+The Hypershift team needs to introduce feature lifecycle into their architecture.
+
+### Risks and Mitigations
+
+There will be a lot of components broken from the start. To see the list of components
+that is going to break, read the "Observed issues with direct OIDC configuration" section
+of the [Unsupported direct use of OIDC as an identity provider](./unsupported-direct-use-of-oidc.md)
+document.
+
+Each of these failures needs to be addressed ad-hoc. Additionally, these are
+only the components we know of. There is a good chance there are third-party
+components that rely on the way OpenShift authentication stack worked until now.
+These will have to read the `authentication/cluster` resource in order to figure
+out how the authentication stack is configured.
+
+### Drawbacks
+
+TODO:
+
+## Design Details
+
+### Test Plan
+
+End to end tests are a MUST for the feature. The tests should check:
+- no kube-apiserver rollout is performed as long as the configuration is invalid
+- the operator handling the configuration API clearly reports failing validation
+- oauth-server and oauth-apiserver are not present when OIDC is configured
+- user can authenticate to the OpenShift API with their ID tokens
+
+We may need manual testing for the following:
+- web console and oauth-proxy work with the new authentication stack
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+A new `KubeAPIServerDirectOIDCConfig` feature gate is introduced. This feature
+gate unlocks the controller configuring the kube-apiserver from reacting to the
+new OIDC API.
+
+The OAuth and User API group objects must be removed by hand by the users prior
+to configuring the direct OIDC identity provider.
+
+Clusters using this feature are prevented from upgrading among major (.y) versions.
```

Comment:
Typo, `.y` is the minor version.

Reply:
I'd consider `.y` a major version in OpenShift context, that's why I specifically outline it here. A minor version would be `.z`, in fact. We allow .z upgrades in our tech preview cluster AFAIK.

Reply:
Thanks for the explanation, wasn't aware of this convention in OCP context :)